MatrixSDK/Crypto/Dehydration/MXDehydrationService.m

```diff
@@ -0,0 +1,295 @@
+            [NSString stringWithFormat:@"curve25519:%@", deviceId]: e2eKeys[@"curve25519"]
+        };
+        deviceInfo.algorithms = [[MXCryptoAlgorithms sharedAlgorithms] supportedAlgorithms];
+        
+        NSString *signature = [account signMessage:[MXCryptoTools canonicalJSONDataForJSON:deviceInfo.signalableJSONDictionary]];
+        deviceInfo.signatures = @{
+                                restClient.credentials.userId: @{
+                                        [NSString stringWithFormat:@"ed25519:%@", deviceInfo.deviceId]: signature
+                                    }
+                                };
+
+        if ([crypto.crossSigning secretIdFromKeyType:MXCrossSigningKeyType.selfSigning])
+        {
+            MXWeakify(self);
+            [crypto.crossSigning signDevice:deviceInfo success:^{
+                MXStrongifyAndReturnIfNil(self);
+                [self uploadDeviceInfo:deviceInfo forAccount:account withMatrixRestClient:restClient success:success failure:failure];
+            } failure:^(NSError * _Nonnull error) {
+                MXLogWarning(@"[MXDehydrationManager] failed to cross-sign dehydrated device data: %@", error);
+                MXStrongifyAndReturnIfNil(self);
+                [self uploadDeviceInfo:deviceInfo forAccount:account withMatrixRestClient:restClient success:success failure:failure];
+            }];
+        } else {
+            [self uploadDeviceInfo:deviceInfo forAccount:account withMatrixRestClient:restClient success:success failure:failure];
+        }
+    } failure:^(NSError *error) {
+        [self stopProgress];
+        MXLogError(@"[MXDehydrationManager] failed to push dehydrated device data: %@", error);
+        dispatch_async(dispatch_get_main_queue(), ^{
+            failure(error);
+        });
+    }];
+}
+
+- (void)rehydrateDeviceWithMatrixRestClient:(MXRestClient*)restClient
+                             dehydrationKey:(NSData*)dehydrationKey
+                                    success:(void (^)(NSString * _Nullable deviceId))success
+                                    failure:(void (^)(NSError *error))failure;
+{
+    MXLogDebug(@"[MXDehydrationManager] rehydrateDevice: getting dehydrated device.");
+    [restClient dehydratedDeviceWithSuccess:^(MXDehydratedDevice *device) {
+        if (!device || !device.deviceId)
+        {
+            MXLogDebug(@"[MXSession] rehydrateDevice: No dehydrated device found.");
+            dispatch_async(dispatch_get_main_queue(), ^{
+                success(nil);
+            });
+            return;
+        }
+        
+        if (![device.algorithm isEqual:MXDehydrationAlgorithm])
+        {
+            MXLogError(@"[MXDehydrationManager] rehydrateDevice: Wrong algorithm for dehydrated device.");
+            dispatch_async(dispatch_get_main_queue(), ^{
+                failure([NSError errorWithDomain:kMXNSErrorDomain code:0 userInfo:@{NSLocalizedDescriptionKey: @"Wrong algorithm for dehydrated device"}]);
```

Comment:
MXDehydrationService should have its own error domain and code like `MXAesErrorDomain` for an objc example.